Abstracts to traits

mecha opened the pull request:
Ports traits from the [`iterator-base`] package, as well as refactors abstract classes into traits.

[`iterator-base`]: https://github.com/Dhii/iterator-abstract

mecha committed:
Ported traits from `iterator-base`

mecha committed:
Refactored `AbstractIterator` into a trait

mecha committed:
Refactored `AbstractRecursiveIterator` to a trait

mecha committed:
Aliased `develop` branch to `0.1.x-dev`

mecha committed:
Added PHPS project files

XedinUnknown commented on src/RecursiveIteratorTrait.php:
```diff
@@ -32,49 +33,98 @@
+        $children = $current->getValue();
+        $this->_pushParent($children);
+
+        if ($this->_isMode(R::MODE_SELF_FIRST)) {
```
Didn't find an implementation of this, which would probably have a contract for `_getMode()` (already exists).

mecha replied:
You're right. Should I remove this and just call `_getMode()` and compare?

XedinUnknown replied:
No, I meant that perhaps there should be an `IsModeCapableTrait`.

mecha replied:
Oh. Don't you think it's simpler to just compare though? I mean, it's an extra trait to implement, test and `use` in classes, all to just compare.

mecha replied:
Then again, determining mode would be abstracted, and not tied to comparing a result from another method (`_getMode()`).

XedinUnknown replied:
Thing is, for most cases you don't need to _know_ the mode; only to _compare_ the mode. Therefore, there may be no "saved" mode, but comparison is done using an algorithm. This is much more flexible.

XedinUnknown reviewed (changes requested):
Masterfully executed in record time.

mecha committed:
Added `IsModeCapableTrait`

XedinUnknown committed:
Added Phan config

However, it isn't currently possible to add `phan/phan` to the
list of required packages because it requires PHP >= 7.0.
Nevertheless, it is still possible to use Phan when it is installed
using other methods.

XedinUnknown commented:
Phan reported the following problems:

```
src\RecursiveIteratorTrait.php:61 PhanTypeMismatchArgument Argument 1 (key) is null but \Dhii\Iterator\RecursiveIteratorTrait::_createIteration() takes \Dhii\Util\String\StringableInterface|string defined at src\RecursiveIteratorTrait.php:305
src\RecursiveIteratorTrait.php:103 PhanTypeMismatchArgument Argument 1 (key) is null but \Dhii\Iterator\RecursiveIteratorTrait::_createIteration() takes \Dhii\Util\String\StringableInterface|string defined at src\RecursiveIteratorTrait.php:305
src\RecursiveIteratorTrait.php:142 PhanTypeMismatchArgument Argument 1 (key) is null but \Dhii\Iterator\RecursiveIteratorTrait::_getElementPathSegment() takes int|string defined at src\RecursiveIteratorTrait.php:342
```

XedinUnknown reviewed (changes requested):
Fix Phan issues

XedinUnknown committed:
Fixing CC config format

XedinUnknown committed:
Removed now redundant autoload bootstrap from Phan config

mecha committed:
Changed irrelevent arg from `null` to empty string

mecha committed:
Fixed param docs

mecha committed:
Merge branch 'task/port-traits-from-base-package' of https://github.com/Dhii/iterator-abstract into task/port-traits-from-base-package

XedinUnknown commented on src/RecursiveIteratorTrait.php:
```diff
@@ -139,7 +139,7 @@ protected function _valid()
     {
         $children = &$this->_getElementChildren($parent);
-        $pathSegment = $this->_getElementPathSegment(null, $parent);
+        $pathSegment = $this->_getElementPathSegment('', $parent);
```
Uh that's not what I meant XD

mecha replied:
I didn't know you meant anything tbh :P

I wasn't comfortable making the `key` param optional for `_getElementPathSegment()`, because that meant the `value` param had to be optional too. But this method makes no sense without parameters.

Then I thought, it's not the fault of the method. It's the fault of the caller. In that context, the key is irrelevant, so any value will do. But empty string makes most sense.

XedinUnknown replied:
Well, yea, the param can't be optional. But it can be `null`. And `null` is an allowed value. So, maybe it can optionally be `null`. Thoughts?

mecha replied:
I was about to say that _that_ wasn't much different than optional param, but then remembered that somewhere it's documented that `null` keys do not show up in the path segment. It's a mechanism for recursive iterators to omit certain "levels" from the path. So yes, allowing `null` keys is the solution.

mecha committed:
Allowing `null` keys in `_getElementPathSegment()`
